osdk/src/bundle.rs

```diff
@@ -0,0 +1,196 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use std::{
+    path::{Path, PathBuf},
+    process::Command,
+};
+
+use crate::{
+    bin::AsterBin,
+    cli::CargoArgs,
+    config_manager::{
+        boot::Boot,
+        qemu::{Qemu, QemuMachine},
+        RunConfig,
+    },
+    error::Errno,
+    error_msg,
+    vm_image::AsterVmImage,
+};
+
+/// The osdk bundle artifact that stores as `bundle` directory.
+///
+/// This `Bundle` struct is used to track a bundle on a filesystem. Every modification to the bundle
+/// would result in file system writes. But the bundle will not be removed from the file system when
+/// the `Bundle` is dropped.
+pub struct Bundle {
+    manifest: BundleManifest,
+    path: PathBuf,
+}
+
+impl Bundle {
+    pub fn new(manifest: BundleManifest, path: impl AsRef<Path>) -> Self {
+        std::fs::create_dir_all(path.as_ref()).unwrap();
+        let created = Self {
+            manifest,
+            path: path.as_ref().to_path_buf(),
+        };
+        created.write_manifest_content();
+        created
+    }
+
+    // FIXME: the load function should be used when implementing build cache, but it is not
+    // implemented yet.
+    #[allow(dead_code)]
+    pub fn load(path: impl AsRef<Path>) -> Self {
+        let manifest_file_path = path.as_ref().join("bundle.toml");
+        let manifest_file_content = std::fs::read_to_string(manifest_file_path).unwrap();
+        let manifest: BundleManifest = toml::from_str(&manifest_file_content).unwrap();
+        // TODO: check integrity of the loaded bundle.
+        Self {
+            manifest,
+            path: path.as_ref().to_path_buf(),
+        }
+    }
+
+    pub fn can_run_with_config(&self, config: &RunConfig) -> bool {
+        // TODO: This pairwise comparison will result in some false negatives. We may
+        // fix it by pondering upon each fields with more care.
+        self.manifest.kcmd_args == config.manifest.kcmd_args
+            && self.manifest.initramfs == config.manifest.initramfs
```

Comment:
The hash value of initramfs should be checked here? Otherwise false positive may occur.